pkg/controller/master/virtualmachine/vmi_controller.go

```diff
@@ -109,3 +110,38 @@ func (h *VMIController) removeDeprecatedFinalizer(_ string, vmi *kubevirtv1.Virt
 	}
 	return vmi, nil
 }
+
+// ReconcileFromHostLabels handles the propagation of metadata from node labels to VirtualMachineInstance annotations.
+func (h *VMIController) PatchMacAddresses(_ string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if vmi == nil || (vmi.DeletionTimestamp == nil && !vmi.IsFinal()) {
+		return vmi, nil
+	}
+
+	vm, err := h.virtualMachineCache.Get(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return vmi, err
+	}
+
+	macAddressStr := vm.Annotations[util.AnnotationMacAddressName]
+	if macAddressStr == "" {
+		return vmi, nil
+	}
+
+	macAddresses := map[string]string{}
+	if err = json.Unmarshal([]byte(macAddressStr), &macAddresses); err != nil {
+		return vmi, err
+	}
+
+	vmCopy := vm.DeepCopy()
+	for i, nic := range vm.Spec.Template.Spec.Domain.Devices.Interfaces {
+		if mac, ok := macAddresses[nic.Name]; ok {
```

Comment:
if vm template already has MAC (e.g. user set it maunally), the annotion should avoid overwrite it